lib/matplotlib/cbook.py

```diff
@@ -2444,3 +2444,15 @@ def _auto_format_str(fmt, value):
         return fmt % (value,)
     except (TypeError, ValueError):
         return fmt.format(value)
+
+
+def _is_pandas_dataframe(x):
+    """Check if 'x' is a Pandas DataFrame."""
+    try:
+        # we're intentionally not attempting to import Pandas. If somebody
+        # has created a Pandas DataFrame, Pandas should already be in sys.modules
+        return isinstance(x, sys.modules['pandas'].DataFrame)
```

Comment:
If we want to raise an exception instead We can do it like so:

```
def _is_pandas_dataframe(x):
    """
    Check if 'x' is a Pandas DataFrame and raise an 
    exception if it is not. If pandas is not loaded,
    raise an exception.
    """
    if 'pandas' not in sys.modules:
        raise ImportError("pandas is not loaded in sys.modules")
    try:
        # we're intentionally not attempting to import Pandas. If somebody
        # has created a Pandas DataFrame, Pandas should already be in sys.modules
        if isinstance(x, sys.modules['pandas'].DataFrame):
            return True
        else:
            raise TypeError("The object is not a pandas.DataFrame object.")
    except Exception as e:
        # Catch all errors
        raise RuntimeError(f"An error occurred: {e}")
```

if pandas is included in the main script and x is not a pandas.DataFrame, this is the traceback:
```
Traceback (most recent call last):
  File "c:\Users\amann\Documents\open source\matplotlib\app.py", line 14, in <module>
    table = ax.table(df, loc='center')
            ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\amann\Documents\open source\matplotlib\lib\matplotlib\table.py", line 750, in table
    if _is_pandas_dataframe(cellText):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\amann\Documents\open source\matplotlib\lib\matplotlib\cbook.py", line 2461, in _is_pandas_dataframe
    raise TypeError("The object is not a pandas.DataFrame object.")
TypeError: The object is not a pandas.DataFrame object.
```

If pandas is not included, this is the traceback:
```
Traceback (most recent call last):
  File "c:\Users\amann\Documents\open source\matplotlib\app.py", line 14, in <module>
    table = ax.table(df, loc='center')
            ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\amann\Documents\open source\matplotlib\lib\matplotlib\table.py", line 750, in table
    if _is_pandas_dataframe(cellText):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\amann\Documents\open source\matplotlib\lib\matplotlib\cbook.py", line 2456, in _is_pandas_dataframe
    raise ImportError("pandas is not loaded in sys.modules")
ImportError: pandas is not loaded in sys.modules
```

Let me know if this is the desired response

Reply:
Please don't change. @story645 I do not follow your argument. `_is_pandas_dataframe` is a substitute for `isinstance(..., pandas.DataFrame)` with the advantage that you don't have to explicitly import pandas first (which doesn't work if pandas is not available). This should only return a bool, and is trivially false if pandas is not installed.

Reply:
This is the same pattern as in `_is_torch_array` and `_is_jax_array`. The pattern was initially suggested here: https://github.com/matplotlib/matplotlib/pull/25887#issuecomment-1910781223

While one should generally catch specific exception, I believe it's better to be general here: If we can't instance-check the object for whatever reason, it's most likely not of the questioned type. When our check fails in an non-anticipated way, we can either say that the object is not of the type (current implementation) or raise (your suggestion). Worst cases:

- Broad exception: It may give the false answer for a user who is supplying the type (but really unlikely) - the user will get a "strange" error further down as it was before introducting the support.
- Narrow exception: If we have overlooked something in the mechanism, it will pop up, very likely also for not-of-the type objects.

The second case may also affect existing users and is far more likely to happen. Therefore, the broad exception is actually defensive here.

If you want to follow up, I suggest to separate this discussion, because this affects also the aforementioned `_is_torch_array` and `_is_jax_array`.